### ISSUE TITLE
Can i load model from JSON format text which dump by using LGBM_BoosterDumpModel

### ISSUE THREAD

**Issue by ShannonBase** [microsoft/LightGBM#6834]
Hi everyone,

As described in title message, I try to load the mode from json text, which was dumped by LGBM_BoosterDumpModel, and it failed. 
The example code as following: 
1: Dump 
```
  int64_t bufflen(1024), out_len{0};
  auto model_content_buff = std::make_unique<char[]>(bufflen);
  memset(model_content_buff.get(), 0x0, bufflen);
  ret =
      LGBM_BoosterDumpModel(booster, 0, -1, C_API_FEATURE_IMPORTANCE_GAIN, bufflen, &out_len, model_content_buff.get());
  if (ret) return nullptr;

  if (out_len > bufflen) {
    model_content_buff.reset(new char[out_len + 1]);
    memset(model_content_buff.get(), 0x0, bufflen);
    ret = LGBM_BoosterDumpModel(booster,
                                0,                              // start iter idx
                                -1,                             // end inter idx
                                C_API_FEATURE_IMPORTANCE_GAIN,  // feature_importance_type
                                out_len,                        // buff len
                                &out_len,                       // out len
                                model_content_buff.get());
  }
```

2: Load
```
  BoosterHandle handle{nullptr};
  int n_iteration;
  auto ret = LGBM_BoosterLoadModelFromString(model_content.c_str(), &n_iteration, &handle);
  return (ret == 0) ? handle : nullptr;
```

the error message outputed by ligthbgm is:
```
 42 [LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.079646 seconds.
 43 You can set `force_row_wise=true` to remove the overhead.
 44 And if memory is not enough, you can set `force_col_wise=true`.
 45 [LightGBM] [Info] Total Bins 702
 46 [LightGBM] [Info] Number of data points in the train set: 32561, number of used features: 14
 47 [LightGBM] [Info] Start training from score 0.240810
 48 2025-02-18T01:45:18.682369Z 10 [System] [MY-015139] [Server] Execution of ALTER TABLE heatwaveml_bench.census_test SECONDARY_LOAD finish    ed. Spent 137 msec for exclusive lock acquire; 2848 msec for secondary engine operation; 75 msec for metadata update and commit.
 49 [LightGBM] [Fatal] Wrong line at model file: {"name": "tree", "version": "v4", "num_class": 1, "objective": "regression", "tree_info": [    {"num_cat": 0, "shrinkage": 1, "num_l
```

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

No, the JSON format cannot currently be used as an input to recreate a Booster. You might find these discussions interesting, and want to click "subscribe" to be notified if there are changes to how model serialization / deserialization is done in LightGBM:

* #4887
* #4681 (especially https://github.com/microsoft/LightGBM/issues/4681#issuecomment-1008181544)
* #4217

You can also look through https://github.com/microsoft/LightGBM/blob/master/src/boosting/gbdt_model_text.cpp if you want to see more details on how model-loading from text files works in LightGBM

**Comment by ShannonBase** [comment]
And another problem,  model type is set to classification, but the output of dump shows the model type is regression.
```
(gdb) p model_content
$8 = "{\"name\":\"tree\",\n\"version\":\"v4\",\n\"num_class\":1,\n\"num_tree_per_iteration\":1,\n\"label_index\":0,\n\"max_feature_idx\":13,\n\"objective\":\"regression\",\n\"average_output\":false,\n\"feature_names\":[\"Column_0\",\"Column_"...
```
the object is `"objective\":\"regression\"`.

the param i used to do train is:
```
  std::string mode_params =
      " task = train "
      " boosting_type = gbdt "
      " objective = binary "
      " metric = binary_logloss,auc "
      " metric_freq = 1 "
      " num_trees = 100 "
      " learning_rate = 0.1 "
      " num_leaves = 63 ";
```

**Comment by jameslamb** [comment]
Hmmm ok, let's see if that's a bug. It will be difficult to investigated this without a **minimal, reproducible example** ([as described here](https://stackoverflow.com/help/minimal-reproducible-example)).

I'm not able to reproduce it in the Python package with LightGBM v4.6.0 (the Python package uses many of these same C API functions).

```python
import lightgbm as lgb
from sklearn.datasets import make_classification

X, y = make_classification()
bst = lgb.train(
    train_set=lgb.Dataset(X, label=y),
    params={
        "objective": "binary",
        "metric": [
            "binary_logloss",
            "auc"
        ],
        "metric_freq": 1,
        "num_trees": 100,
        "learning_rate": 0.1,
        "num_leaves": 63,
        "task": "train"
    }
)
model_json = bst.dump_model()

model_json["objective"]
# "binary sigmoid:1"
```

* what version of LightGBM are you using?
* what operating system?
* can you share some logs from training?
* are you able to provide a minimal, reproducible example?

**Comment by ShannonBase** [comment]
> Thanks for using LightGBM.
> 
> No, the JSON format cannot currently be used as an input to recreate a Booster. You might find these discussions interesting, and want to click "subscribe" to be notified if there are changes to how model serialization / deserialization is done in LightGBM:
> 
> * [[RFC] Unify model format customize string or Json #4887](https://github.com/microsoft/LightGBM/issues/4887)
> * [Serializers convert floats to strings #4681](https://github.com/microsoft/LightGBM/issues/4681) (especially [Serializers convert floats to strings #4681 (comment)](https://github.com/microsoft/LightGBM/issues/4681#issuecomment-1008181544))
> * [Saving Model to binary stream #4217](https://github.com/microsoft/LightGBM/issues/4217)
> 
> You can also look through https://github.com/microsoft/LightGBM/blob/master/src/boosting/gbdt_model_text.cpp if you want to see more details on how model-loading from text files works in LightGBM

OK. I will change my impl, and waiting for your team works on unifying the model format.

**Comment by ShannonBase** [comment]
> Hmmm ok, let's see if that's a bug. It will be difficult to investigated this without a **minimal, reproducible example** ([as described here](https://stackoverflow.com/help/minimal-reproducible-example)).
> 
> I'm not able to reproduce it in the Python package with LightGBM v4.6.0 (the Python package uses many of these same C API functions).
> 
> import lightgbm as lgb
> from sklearn.datasets import make_classification
> 
> X, y = make_classification()
> bst = lgb.train(
>     train_set=lgb.Dataset(X, label=y),
>     params={
>         "objective": "binary",
>         "metric": [
>             "binary_logloss",
>             "auc"
>         ],
>         "metric_freq": 1,
>         "num_trees": 100,
>         "learning_rate": 0.1,
>         "num_leaves": 63,
>         "task": "train"
>     }
> )
> model_json = bst.dump_model()
> 
> model_json["objective"]
> # "binary sigmoid:1"
> * what version of LightGBM are you using?
> * what operating system?
> * can you share some logs from training?
> * are you able to provide a minimal, reproducible example?

The version i used:
```
cat VERSION.txt 
4.5.0
```
OS:
```
uname -a
Linux xxx-virtual-machine 6.8.0-51-generic #52-Ubuntu SMP PREEMPT_DYNAMIC Thu Dec  5 13:09:44 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

example code:
Train:
```
  std::string parameters = "max_bin=254 ";
  DatasetHandle train_dataset_handler{nullptr};
  //clang-format off
  LGBM_DatasetCreateFromMat(training_data,       // feature data ptr
                                       data_type,           // type of sampe data
                                       n_data,              // # of sample data
                                       n_feature,           // # of features
                                       1,                   // 1 row-first, 0 column-first.
                                       parameters.c_str(),  // params.
                                       nullptr,             // ref.
                                       &train_dataset_handler);

  // to set label data
  LGBM_DatasetSetField(train_dataset_handler, "label", label_data, n_data, label_data_type);
  //clang-format on
  
  BoosterHandle booster;
  LGBM_BoosterCreate(train_dataset_handler, task_mode.c_str(), &booster);
  LGBM_BoosterAddValidData(booster, train_dataset_handler);
  
  int finished{0};
  for (auto iter = 0; iter < 100; ++iter) {
    LGBM_BoosterUpdateOneIter(booster, &finished);
    if (finished) break;
  }

  int64_t bufflen(1024), out_len{0};
  auto model_content_buff = std::make_unique<char[]>(bufflen);
  memset(model_content_buff.get(), 0x0, bufflen);
  LGBM_BoosterDumpModel(booster, 0, -1, C_API_FEATURE_IMPORTANCE_GAIN, bufflen, &out_len, model_content_buff.get());
  
  if (out_len > bufflen) {
    model_content_buff.reset(new char[out_len + 1]);
    memset(model_content_buff.get(), 0x0, bufflen);
    LGBM_BoosterDumpModel(booster,
                                0,                              // start iter idx
                                -1,                             // end inter idx
                                C_API_FEATURE_IMPORTANCE_GAIN,  // feature_importance_type
                                out_len,                        // buff len
                                &out_len,                       // out len
                                model_content_buff.get());
  }

  LGBM_DatasetFree(train_dataset_handler);
  LGBM_BoosterFree(booster);
  model_content.assign(model_content_buff.get(), out_len);
```
data type: `C_API_DTYPE_FLOAT64`, label data type: `C_API_DTYPE_FLOAT32`

**Comment by jameslamb** [comment]
Thanks! So, that is different from what you claimed previously. Previously, you said you used these parameters:

```cpp
  std::string mode_params =
      " task = train "
      " boosting_type = gbdt "
      " objective = binary "
      " metric = binary_logloss,auc "
      " metric_freq = 1 "
      " num_trees = 100 "
      " learning_rate = 0.1 "
      " num_leaves = 63 ";
```

But the code you just provided is instead doing this:

```text
std::string parameters = "max_bin=254 ";
```

If you do not explicitly provide `objective` in parameters, LightGBM defaults to `"regression"`.

https://github.com/microsoft/LightGBM/blob/3fad53bc587acc634592ee168aa59ca21cba83f4/include/LightGBM/config.h#L165

**Comment by ShannonBase** [comment]
> Thanks! So, that is different from what you claimed previously. Previously, you said you used these parameters:
> 
>   std::string mode_params =
>       " task = train "
>       " boosting_type = gbdt "
>       " objective = binary "
>       " metric = binary_logloss,auc "
>       " metric_freq = 1 "
>       " num_trees = 100 "
>       " learning_rate = 0.1 "
>       " num_leaves = 63 ";
> But the code you just provided is instead doing this:
> 
> ```
> std::string parameters = "max_bin=254 ";
> ```
> 
> If you do not explicitly provide `objective` in parameters, LightGBM defaults to `"regression"`.
> 
> [LightGBM/include/LightGBM/config.h](https://github.com/microsoft/LightGBM/blob/3fad53bc587acc634592ee168aa59ca21cba83f4/include/LightGBM/config.h#L165)
> 
> Line 165 in [3fad53b](/microsoft/LightGBM/commit/3fad53bc587acc634592ee168aa59ca21cba83f4)
> 
>  std::string objective = "regression";

Thanks, I.C. Now, close it. Waiting for works on unify mode format. ;-)
